src/gprs.c

```diff
@@ -3186,21 +3192,46 @@ static void gprs_load_settings(struct ofono_gprs *gprs, const char *imsi)
 static struct pri_context *gprs_context_for_ia(struct ofono_gprs *gprs)
 {
 	GSList *l;
+	struct pri_context *ctx_ia = NULL;
+	struct pri_context *ctx_inet_pref = NULL;
+	struct pri_context *ctx_inet = NULL;
+	struct pri_context *ctx_other_pref = NULL;
 
 	for (l = gprs->contexts; l; l = l->next) {
 		struct pri_context *ctx = l->data;
 
-		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA)
-			return ctx;
-	}
+		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA) {
+			if (ctx->preferred)
+				return ctx;
 
-	for (l = gprs->contexts; l; l = l->next) {
-		struct pri_context *ctx = l->data;
+			if (ctx_ia == NULL)
+				ctx_ia = ctx;
+		} else if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_INTERNET) {
 
-		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_INTERNET)
-			return ctx;
+			if (ctx->preferred && ctx_inet_pref == NULL)
+				ctx_inet_pref = ctx;
+			else if (ctx_inet == NULL)
+				ctx_inet = ctx;
+
+		} else if (ctx->preferred && ctx_other_pref == NULL) {
+			ctx_other_pref = ctx;
+		}
+	}
+
+	if (ctx_ia != NULL) {
+		set_preferred(ctx_ia, ofono_dbus_get_connection(), TRUE);
+		return ctx_ia;
 	}
 
+	if (ctx_inet_pref != NULL)
+		return ctx_inet_pref;
+
+	if (ctx_inet != NULL)
+		return ctx_inet;
+
+	if (ctx_other_pref != NULL)
+		return ctx_other_pref;
+
 	return gprs->contexts->data;
```

Comment:
I think you need to add a check to ensure that gprs->contexts isn't NULL before this final return statement.  I know the default behavior of ofono is to provision a single Internet context if none are found, however it's possible that the user could delete this APN.  The loop that iterates over the contexts correctly checks it before using, so the same should be done here.